src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -601,28 +601,20 @@ internal MemberAccessor MemberAccessorStrategy
 
         /// <summary>
         /// Initializes the converters for the reflection-based serializer.
-        /// <seealso cref="InitializeForReflectionSerializer"/> must be checked before calling.
         /// </summary>
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        internal void InitializeForReflectionSerializer()
+        [MemberNotNull(nameof(s_reflectionTypeInfoResolver))]
+        internal static void EnsureInitializedForReflectionSerializer()
         {
-            RootReflectionSerializerDependencies();
-            Volatile.Write(ref IsInitializedForReflectionSerializer, true);
-            if (_cachingContext != null)
-            {
-                _cachingContext.Options.IsInitializedForReflectionSerializer = true;
-            }
+            s_reflectionTypeInfoResolver ??= new DefaultJsonTypeInfoResolver();
         }
 
         private JsonTypeInfo GetJsonTypeInfoFromContextOrCreate(Type type)
         {
             JsonTypeInfo? info = _serializerContext?.GetTypeInfo(type);
-            if (info == null && IsInitializedForReflectionSerializer)
+            if (info == null && s_reflectionTypeInfoResolver != null)
```

Comment:
This will cause every source generating serializer in the process to fall back to reflection just because some unrelated thread ran one reflection-based serialization. Please revert this change.